### PR TITLE
Fix and improve eval metrics logging

### PR DESCRIPTION
**Pull request by aireenmei**
# Description

b/509626795
### Summary of issue:
When running eval after every train step (eval_interval=1), it's observed that train step time in the logging increases significantly, from ~9.5s -> 11.5s

### Root cause analysis:
* Major: in the current code, the train step time boundary mistakenly includes the eval loop. So the train step time right before eval includes eval time, causing step time inflation. In most cases people use a bigger eval_interval and only see the inflated step once in many steps and ignore the issue.
* Minor: Unlike train step time that always delay reporting the metrics one step to avoid blocking the data loading of next step. In eval, the metrics are calculated right after the last eval step (the only eval step in this case), blocking the data loading of the following training step.

### Changes in this PR
* Fix the train/eval step time boundary. Train and eval step times are separated and print out clearly.
* Use the same buffer mechanism for train and eval metrics. Train and eval metrics share the same buffer. When the new step metrics comes in, print out the previous step metrics in the buffer.
* Improve eval metrics observability by printing out eval running average metrics every eval step
* before the PR, eval metrics look like:
```
completed step: k, seconds, ...., loss:    # train step
Completed eval step 0  # no time span or metrics
Completed eval step 1
....
eval metrics after step: k, loss= ....  # after finished all eval steps, print avg metrics of all evaluated data
```
* With this PR, eval metrics look like:
```
completed step: k, seconds, ...., loss:    # train step
Starting eval after train step k
completed eval step: 0,  seconds: x, running loss=xxx  # with time span and running metrics
completed eval step: 1,  seconds: x, running loss=xxx  # running loss is avg loss of all eval data seen so far
....
Completed eval after train step k, loss=xxx, ...  # loss is avg loss of all evaluated data
```

# Tests

Test on v5e-32, default 1b model, per_device_batch_size=1
* With the old code, step time ~ 1.6s, [screenshot](https://screenshot.googleplex.com/AfMXUX7kwnNtcnk)
* With the new code, step time separated, train ~1s + eval ~0.5s, [screenshot](https://screenshot.googleplex.com/5ByLY8zXvurYkok). There is slightly in the total time likely due to fixing the minor blocking issue stated above. 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/common/metric_logger.py [hunk 1]
```diff
@@ -324,19 +356,68 @@ def get_performance_metric_queue(self, config):
...
+    Both train and eval entries carry raw metrics and defer all processing
+    (record_train_metrics / _accumulate_eval_metrics) to _flush_one_buffered_entry so float()
+    never blocks the dispatch path.
+    """
+    if self.buffered_metrics:
+      self._flush_one_buffered_entry(self.buffered_metrics.pop(0))
+    if is_training:
+      if self._pending_eval_step_count > 0:
+        self._finalize_eval_metrics(step)
+      self.buffered_metrics.append(("train", step, metrics, step_time_delta))
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply src/maxtext/common/metric_logger.py]: done

File: src/maxtext/experimental/rl/grpo_trainer.py [hunk 2]
```diff
@@ -877,27 +877,32 @@ def generation_worker_fn(
...
           eval_step_count += 1
-        metric_logger.record_eval_metrics(step, eval_step_count=eval_step_count)
-        if metric_logger.cumulative_eval_metrics["scalar"]["eval/avg_loss"] <= config.target_eval_loss:
-          prof.deactivate()
-          raise exceptions.StopTraining(f"Target loss {config.target_eval_loss=} is achieved.")
 
       prof.maybe_deactivate_profiler(step, state)
 
       if step == start_step:
         max_utils.print_mem_stats("After params initialized")
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply src/maxtext/experimental/rl/grpo_trainer.py:902]: done

File: src/maxtext/trainers/post_train/sft/train_sft_deprecated.py [hunk 3]
```diff
@@ -124,27 +124,32 @@ def train_loop(config, recorder, state=None):
...
+              eval_metrics, eval_step_count, step_time_delta=eval_step_time_delta, is_training=False
+          )
           max_logging.log(f"Completed eval step {eval_step_count}")
           eval_step_count += 1
-        metric_logger.record_eval_metrics(step, eval_step_count=eval_step_count)
-        if metric_logger.cumulative_eval_metrics["scalar"]["eval/avg_loss"] <= config.target_eval_loss:
-          prof.deactivate()
-          raise exceptions.StopTraining(f"Target loss {config.target_eval_loss=} is achieved.")
 
       prof.maybe_deactivate_profiler(step, state)
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply src/maxtext/trainers/post_train/sft/train_sft_deprecated.py:147]: done

File: src/maxtext/common/metric_logger.py [hunk 4]
```diff
@@ -324,19 +356,68 @@ def get_performance_metric_queue(self, config):
...
+      self.write_metrics(metrics, step)
+    elif kind == "eval":
+      _, eval_step, raw_metrics, step_time_delta = entry
+      # _accumulate_eval_metrics calls float() that materialize the metrics, deferred to here
+      self._accumulate_eval_metrics(raw_metrics)
+      running_count = eval_step + 1  # eval_step is 0-indexed
+      cumulative = self.cumulative_eval_metrics["scalar"]
+      running_avg_loss = cumulative["eval/total_loss"] / (cumulative["eval/total_weights"] + EPS)
+      snapshot = {
+          "scalar": {
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply src/maxtext/common/metric_logger.py:394]: Will remove running_eval from tensorboard logging

[the remark this answers is not in the document]
Comment (aireenmei) [reply src/maxtext/common/metric_logger.py:394]: done

**Review by RissyRan — APPROVED** [review]
Thanks Aireen for the change!

**Review by shuningjin — APPROVED** [review]
